File: code/1510. Stone Game IV.py
class Solution:
    def winnerSquareGame(self, n: int) -> bool:
        mp = dict()
        def helper(n):
            if n in mp:
                return mp[n]
            for i in range(int(math.sqrt(n)), 0, -1):
                if not helper(n - i * i):
                    mp[n] = True
                    return mp[n]
            mp[n] = False
            return mp[n]
        return helper(n)

File: code/1510. Stone Game IV.cpp
class Solution {
public:
    bool winnerSquareGame(int n) {
        unordered_map<int, bool> mp;
        function<bool(int)> helper = [&](int stones) {
            if (mp.count(stones)) {
                return mp[stones];
            }
            for (int i = sqrt(stones); i >= 1; i--) {
                if (!helper(stones - i * i)) {
                    return mp[stones] = true;
                }
            }
            return mp[stones] = false;
        };
        return helper(n);
    }
};